cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -1228,43 +1232,53 @@ public void setInstances(ProgramId programId, int instances) throws Exception {
    * @return the programs in the provided namespace
    */
   public List<ProgramRecord> list(NamespaceId namespaceId, ProgramType type) throws Exception {
-    Collection<ApplicationSpecification> appSpecs = store.getAllApplications(namespaceId);
-    List<ProgramRecord> programRecords = new ArrayList<>();
-    for (ApplicationSpecification appSpec : appSpecs) {
-      switch (type) {
-        case MAPREDUCE:
-          createProgramRecords(namespaceId, appSpec.getName(), type, appSpec.getMapReduce().values(), programRecords);
-          break;
-        case SPARK:
-          createProgramRecords(namespaceId, appSpec.getName(), type, appSpec.getSpark().values(), programRecords);
-          break;
-        case SERVICE:
-          createProgramRecords(namespaceId, appSpec.getName(), type, appSpec.getServices().values(), programRecords);
-          break;
-        case WORKER:
-          createProgramRecords(namespaceId, appSpec.getName(), type, appSpec.getWorkers().values(), programRecords);
-          break;
-        case WORKFLOW:
-          createProgramRecords(namespaceId, appSpec.getName(), type, appSpec.getWorkflows().values(), programRecords);
-          break;
-        default:
-          throw new Exception("Unknown program type: " + type.name());
-      }
+    Function<ApplicationSpecification, Iterable<? extends ProgramSpecification>> func;
+    switch (type) {
+      case MAPREDUCE:
+        func = appSpec -> appSpec.getMapReduce().values();
```

Comment:
nit: you can use method reference, e.g. `appSpec.getMapReduce()::values` instead of lambda